planner/core/optimizer.go

```diff
@@ -687,31 +695,186 @@ const (
 	unknown shuffleTarget = iota
 	window
 	joinBuild
+	hashAgg
 )
 
 type fineGrainedShuffleHelper struct {
 	shuffleTarget shuffleTarget
 	plans         []*basePhysicalPlan
+	joinKeysCount int
+}
+
+type tiflashClusterInfoStatus uint8
+
+const (
+	unInitialized tiflashClusterInfoStatus = iota
+	initialized
+	failed
+)
+
+type tiflashClusterInfo struct {
+	itemStatus tiflashClusterInfoStatus
+	itemValue  uint64
 }
 
 func (h *fineGrainedShuffleHelper) clear() {
 	h.shuffleTarget = unknown
 	h.plans = h.plans[:0]
+	h.joinKeysCount = 0
 }
 
 func (h *fineGrainedShuffleHelper) updateTarget(t shuffleTarget, p *basePhysicalPlan) {
 	h.shuffleTarget = t
 	h.plans = append(h.plans, p)
 }
 
-func setupFineGrainedShuffleInternal(plan PhysicalPlan, helper *fineGrainedShuffleHelper, streamCount uint64) {
+// calculateTiFlashStreamCountUsingMinLogicalCores uses minimal logical cpu cores among tiflash servers, and divide by 2
+// return false, 0 if any err happens
+func calculateTiFlashStreamCountUsingMinLogicalCores(ctx context.Context, sctx sessionctx.Context, serversInfo []infoschema.ServerInfo) (bool, uint64) {
+	failpoint.Inject("mockTiFlashStreamCountUsingMinLogicalCores", func(val failpoint.Value) {
+		intVal, err := strconv.Atoi(val.(string))
+		if err == nil {
+			failpoint.Return(true, uint64(intVal))
+		} else {
+			failpoint.Return(false, 0)
+		}
+	})
+	rows, err := infoschema.FetchClusterServerInfoWithoutPrivilegeCheck(ctx, sctx, serversInfo, diagnosticspb.ServerInfoType_HardwareInfo, false)
+	if err != nil {
+		return false, 0
+	}
+	var initialMaxCores uint64 = 10000
+	var minLogicalCores uint64 = initialMaxCores // set to a large enough value here
+	for _, row := range rows {
+		if row[4].GetString() == "cpu-logical-cores" {
+			logicalCpus, err := strconv.Atoi(row[5].GetString())
+			if err == nil && logicalCpus > 0 && uint64(logicalCpus) < minLogicalCores {
+				minLogicalCores = uint64(logicalCpus)
+			}
+		}
+	}
+	// No need to check len(serersInfo) == serverCount here, since missing some servers' info won't affect the correctness
+	if minLogicalCores > 1 && minLogicalCores != initialMaxCores {
+		return true, minLogicalCores / 2
```

Comment:
Just question: why divide by 2?

Reply:
Just copied from our TiFlash maxStream calculation: use actual available physical cores. And since no info can found here to infer the ratio of logical/physical, just use the most common 2 here.